### PR TITLE
Support foregroundtask registration for packaged applications

### PR DESCRIPTION
**Pull request by sharath2727**

### CODE REVIEW

File: test/TestApps/PushNotificationsDemoPackage/Package.appxmanifest
Comment (loneursid) [inline test/TestApps/PushNotificationsDemoPackage/Package.appxmanifest:15]: Why the bump in version number? Is this intentional?

File: test/TestApps/PushNotificationsDemoApp/main.cpp
Comment (loneursid) [inline test/TestApps/PushNotificationsDemoApp/main.cpp:21]: Why the change in the guid? is this intentional or required?

File: test/TestApps/PushNotificationsDemoPackage/PushNotificationsDemoPackage.wapproj
Comment (pmpurifoy) [inline test/TestApps/PushNotificationsDemoPackage/PushNotificationsDemoPackage.wapproj:51]: Make sure you cleanup local changes 😄

File: dev/PushNotifications/PushNotificationChannel.h
Comment (pmpurifoy) [inline dev/PushNotifications/PushNotificationChannel.h:38]: In this scenario, when the foreground isn't handled, we need to find a way to route the args to background activation. So that means protocol activation will have to happen here 😄 . Check the LRP to see how protocol activation is currently being done. Maybe start a commons file shared between LRP/WAS.

File: test/TestApps/PushNotificationsDemoPackage/PushNotificationsDemoPackage.wapproj
Comment (hulumane) [reply test/TestApps/PushNotificationsDemoPackage/PushNotificationsDemoPackage.wapproj:51]: This should be a relative path I imagine.

File: dev/PushNotifications/PushNotificationChannel.h
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.h:34]: We probably should file a bug here to use the correlation vector in our telemetry?

File: dev/PushNotifications/PushNotificationChannel.h
Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.h:38]: Correct. If the foreground event is not handled, we should re-launch the app with the payload using Protocol. That would mean that this line of code needs to go away:
 THROW_HR_IF(E_NOT_VALID_STATE, !foregroundHandled);

Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.h:38]: Also please test this corner case scenario : )

File: test/TestApps/PushNotificationsDemoPackage/PushNotificationsDemoPackage.wapproj
Comment (sharath2727) [reply test/TestApps/PushNotificationsDemoPackage/PushNotificationsDemoPackage.wapproj:51]: @pmpurifoy that is why still in draft mode :)

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:108]: This is incorrect. It should be:
THROW_HR_IF_NULL(E_UNEXPECTED, m_channelInfo);

You essentially want to return an error if channelInfo is not set, right? So use the winrt reference, not the fields in the reference.

File: dev/PushNotifications/PushNotificationChannel.h
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.h:38]: I agree to the fact that we should not have the line set.  Firslty, let me reiterate we are dealing with packaged applications. Let me put the scenarios here:

1. Foreground handled - args.Cancel is set to true - this will avoid background task to be triggered
2. Foreground not handled - args.Cancel is set to false - then if the background task is also registered by the user then backgroundtask is triggered.

The above cases are handled if we remove the foregroundhandled check. Do we really need the protocol activation here.?

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:118]: Nit: Too many blank lines. Please remove them all before the brace.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:121]: Remove comment.

File: dev/PushNotifications/PushNotificationChannel.h
Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.h:38]: I'm not sure I understand the question. The contract is straightforward here: If the developer chooses not to intercept the foreground event, then the expectation is that the background trigger is executed. In this case, there is a noop. So it is a bug : )

Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.h:38]: I also think that packaged apps using BI suffer the same limitation here. We should file a seperate bug for that and fix it as well.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (robertzhou-wpn) [inline dev/PushNotifications/PushNotificationChannel.cpp:155]: Nit: looks to me to register and unregister a sink, we will first get the appUserModelId. Can we put the logic of getting the appUserModelId into a function, which can then be called in both places?

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.cpp:108]: Yeah I am aware of this. I have taken a note. Will update the PR :)

File: dev/PushNotifications/PushNotificationChannel.h
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.h:34]: Yeah I created one: https://github.com/microsoft/WindowsAppSDK/issues/1464

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.cpp:108]: Actually m_channelInfo is a struct static allocation object and not a pointer object. So checking on the channelUri string should be the right thing to do?

File: test/TestApps/PushNotificationsDemoApp/main.cpp
Comment (sharath2727) [reply test/TestApps/PushNotificationsDemoApp/main.cpp:21]: Planning to change the GUID as the newer guid works for both scenarios - packaged and unpackaged.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.cpp:107]: Check out std::variant here. https://en.cppreference.com/w/cpp/utility/variant. You could technically use this and check if it is invalid or not?

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:114]: No need of having a token local variable here. Just return this at the end.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:112]: You only want to register the foreground sink if the foregroundHandlerCount is 0. Otherwise, we keep the existing sink.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:153]: We only unregister the sink if the foregroundHandlerCount == 0

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:188]: try catch. Don't really understand this flow at all.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:197]: This is all shared code between the LRP and WinSDK. So it should be defined in a common header in a utility function.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:205]: Did you look into @danielayala94 's code change for special characters and protocol launch? I think we are introducing the same bug here which he fixed. Hence its important to share code through utility helpers so that we don't redo the same mistakes.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:191]: foregroundHandled should be used along with channelInfo != nullptr to determine protocol launch

File: dev/PushNotifications/externs.h
Comment (hulumane) [inline dev/PushNotifications/externs.h:36]: @danielayala94 Already introduced a function for this. We should not copy paste but introduce Utility Helpers with shared common code.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:213]: Did you test the case where the foregroundHandled = false? We should ensure that a new background process is launched with the payload.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.cpp:107]: BTW c_str() can never be null for a string. It can only be empty.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.cpp:205]: Agreed. I totally missed that. I have integrated that change into my code and validated it. I have created a helper for the same. Thanks.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.cpp]: question here: should we not always register the latest sink? @hulumane.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.cpp]: For now I have moved this code into externs.h which the files in PushNotifications/ are using. I need to see how I can make both LRP and WinSDK use the same code when daniel's changes and this PR merges into WNP. It makes my life easier then. I will take this as a bug: https://github.com/microsoft/WindowsAppSDK/issues/1481

Thanks.

File: dev/PushNotifications/externs.h
Comment (sharath2727) [reply dev/PushNotifications/externs.h]: Currently moved all this into externs.h. Will create a utility to be used between LRP and PushNotifications. bug: https://github.com/microsoft/WindowsAppSDK/issues/1481

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [reply dev/PushNotifications/PushNotificationChannel.cpp]: You misunderstand the design for foreground sink. The way @pmpurifoy  implemented this is using a single sink registration which is refcounted for every eventHandler that the developer subscribes too. 

Another way to achieve the same result is to AddRef the COM ref instead of having a foregroundHandlerCount (we can delete this). That might actually be more optimal.

Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.cpp]: https://github.com/microsoft/WindowsAppSDK/issues/1489 - created a bug for this as this is an improvement and will not impact the current design. Thanks.

File: dev/PushNotifications/PushNotificationChannel.h
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationChannel.h]: https://github.com/microsoft/WindowsAppSDK/issues/1490

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (danielayala94) [inline dev/PushNotifications/PushNotificationUtility.cpp:1]: Should the utility implementation be in the header file? I would imagine that this file has to be published to be consumed by the LRP. But also, if the LRP compiles before the SDK, wouldn't a compile error happen since we were looking for the utility file not having published yet?

Would dev/Common be the right place for this code?

File: dev/PushNotifications/PushNotificationUtility.h
Comment (danielayala94) [inline dev/PushNotifications/PushNotificationUtility.h:7]: Nit:

```suggestion
const std::wstring ConvertByteArrayToWideString(unsigned int payloadLength, _In_reads_(payloadLength) byte* payload) noexcept;
```

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (danielayala94) [inline dev/PushNotifications/PushNotificationUtility.cpp:31]: This can't be noexcept. For instance, you are dealing with STL strings.

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (danielayala94) [inline dev/PushNotifications/PushNotificationUtility.cpp:26]: THROW but noexcept?

**Review by danielayala94 — CHANGES_REQUESTED** [review]

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationUtility.cpp:1]: Create a common utility file between LRP and PushNotifications #1481 - That is the reason I have created this bug. Temporarily I have created the files PushNotificationUtility.h/cpp in dev/PushNotifications. As responses might be little longer it should not block the PR from merging atleast in WNP_LRP. By then once we know where these files should move to we can address it as part of a bug. Let me know if you think otherwise. Thanks.

Comment (sharath2727) [reply dev/PushNotifications/PushNotificationUtility.cpp:1]: @danielayala94 - realized that PushNotifications.vcxitems is a shared folder. So we don't need to create a new common folder and with the suggestions from Howard, we should be able to call into PushNotificationUtility.h/cpp from anywhere within LRP and beyond. Thanks.

File: dev/PushNotifications/PushNotificationUtility.h
Comment (jonwis) [inline dev/PushNotifications/PushNotificationUtility.h:10]: Consider having this method return the `wil::unique_cotaskmem_string` instead.

File: dev/PushNotifications/PushNotificationUtility.h
Comment (jonwis) [inline dev/PushNotifications/PushNotificationUtility.h:15]: Presumably this should check for the one "no package identity" error and return null, then throw otherwise?

**Review by jonwis — CHANGES_REQUESTED** [review]
Two things to change, the rest are optional - use of that "count" thing to track number of registered handlers, and the missing function-level try/catch on the noexcept method. Otherwise looks pretty good, glad to see this coming together.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (jonwis) [reply dev/PushNotifications/PushNotificationChannel.cpp:108]: ... or if that's too hard, put the `Count++` _after_ you've registered and added the handler.

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (robertzhou-wpn) [inline dev/PushNotifications/PushNotificationChannel.cpp:148]: Looks like the counter here will have the similar problem as Jon mentioned in the counter adding case. can we move the count operation after the operations?

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationUtility.cpp:34]: This is common code pasted from the LRP. Can we use a utility helper method here?

File: dev/PushNotifications/PushNotificationUtility.h
Comment (hulumane) [inline dev/PushNotifications/PushNotificationUtility.h:17]: I believe this throws for the no memory case. But your helper method doesn't handle that.

File: dev/PushNotifications/PushNotificationUtility.h
Comment (hulumane) [inline dev/PushNotifications/PushNotificationUtility.h:15]: If this API call throws, then the caller method should wrap a try catch_return. If the API returns HRESULT, then we need a RETURN_IF_FAILED here or a THROW_IF_FAILED with a method level try-catch_return()

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:104]: Comment needed here that this else is specifically for the RemoteId case when we directly call into the UDK from the SDK to requestChannel from RemoteId

File: dev/PushNotifications/PushNotificationChannel.cpp
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.cpp:187]: Please set this to true instead. Default behavior should be that if the dev doesn't set this to true, then we don't do protocol launch.

File: dev/PushNotifications/PushNotificationChannel.h
Comment (hulumane) [inline dev/PushNotifications/PushNotificationChannel.h:33]: Nit: Missing space
// INotificationHandler

File: dev/PushNotifications/PushNotificationUtility.h
Comment (hulumane) [reply dev/PushNotifications/PushNotificationUtility.h:15]: I thibk this is missing a THROW_IF_FAILED @sharath2727 ?

Comment (jonwis) [reply dev/PushNotifications/PushNotificationUtility.h:15]: It's OK to have it return null, but the caller has to know that through a comment in the header - "method returns nullptr when the package does not have identity."

Comment (sharath2727) [reply dev/PushNotifications/PushNotificationUtility.h:15]: Sure. I think throwing on the error codes: https://docs.microsoft.com/en-us/windows/win32/api/appmodel/nf-appmodel-getcurrentapplicationusermodelid should take care of this.

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (aeloros) [reply dev/PushNotifications/PushNotificationUtility.cpp:38]: I can't remember if we talked about this before or not,  but I believe we would be better served not creating some new payload model.  Instead we should create a contractId for the push server, and use the already existing infrastructure to pass properties through the query string.  This might be an example of a fragile design since it requires continued upkeep to make it special.

**Comment by loneursid** [comment]
So, I've messed up. I've acidentaly merged this branch into WNP_LRP. As a result GIT closed this PR and the branch was deleted. I've reverted the merge from WNP_LRP and restored the branch. Still trying to figure out how to re-open this PR.

**Review by jonwis — COMMENTED** [review]
Hmm

File: dev/PushNotifications/PushNotificationUtility.cpp
Comment (sharath2727) [reply dev/PushNotifications/PushNotificationUtility.cpp]: @jonwis /@aeloros - I see howard introducing a lot of string conversion APIs in dev/common: https://github.com/microsoft/WindowsAppSDK/pull/1513. Once the PR merges some of these conversion APIs in my .h/.cpp will be removed. 

@aeloros - "Instead we should create a contractId for the push server, and use the already existing infrastructure to pass properties through the query string. This might be an example of a fragile design since it requires continued upkeep to make it special" -  can you please help me understand this further? Maybe I could work on this going ahead to simplify things here.

Comment (aeloros) [reply dev/PushNotifications/PushNotificationUtility.cpp]: Sure, the command-line on the receiving end (when GetActivatedEventArgs() parses it) already has a format that supports passing payload information.  Specifically I'm talking about the ability to pass a Uri that contains a key/value pair (the query string).

There are problems with the current implementation of GetActivatedEventArgs() in that it doesn't tolerate malformed command-lines well.  I have a change out to fix that, but it changes the expectations of the format as well.  

I'm mostly suggesting here that we should try to find a better way to utilize the command-line.  One that doesn't have the calling developer designing custom formats to support payloads.  We should instead find a way to utilize the Uri format and serialize + encode data into that format.  

This also lends itself to creating shared functions in the code base that generate that command-line.  I've started to do that in a super simple way with GenerateCommandLine() inside of ActivationRegistrationManager.cpp.  It could be moved and enhanced or accompanied with other helper functions to make a useful reusable toolset for callers to create command-lines in a consistent way.

I don't think any of this is the end of the world if not done.  At this point in the conversation though, it probably would be intentional technical debt.  Obviously we should avoid that when possible, but dealing with technical debt isn't the only factor in our lives. :)

**Pull request merged**
